### PR TITLE
TCP: add support for SYN cookies

### PR DESCRIPTION
**Pull request by francescolavra**
This allows a TCP server to stay responsive when subject to "SYN flood" DoS attacks, without allocating memory for each SYN packet received from the attacker.
The hash function being used to calculate SYN cookie values is MD5.

### CODE REVIEW

File: src/core/tcp.c [hunk 1]
```diff
@@ -203,6 +203,10 @@ tcp_init(void)
 #ifdef LWIP_RAND
   tcp_port = TCP_ENSURE_LOCAL_PORT_RANGE(LWIP_RAND());
 #endif /* LWIP_RAND */
+#if TCP_LISTEN_BACKLOG
+  for (int i = 0; i < TCP_SYNCOOKIE_SECRET_SIZE / sizeof(u32_t); i++)
+  *(u32_t *)(tcp_syncookie_secret + i * sizeof(u32_t)) = LWIP_RAND();
```
Comment (wjhun) [inline src/core/tcp.c]: Total cosmetic nit, but casting tcp_syncookie_secret first and then adding avoids multiply. Also indent.

File: src/core/tcp.c [hunk 2]
```diff
@@ -324,6 +328,8 @@ tcp_backlog_accepted(struct tcp_pcb *pcb)
       pcb->listener->accepts_pending--;
       tcp_clear_flags(pcb, TF_BACKLOGPEND);
     }
+  } else if ((pcb->state == SYN_RCVD) && (pcb->listener != NULL)) {
+    pcb->listener->syn_pending--;
```
Comment (wjhun) [inline src/core/tcp.c:333]: LWIP_ASSERT(pcb->listener->syn_pending != 0) would be nice, similar to what is done elsewhere with accepts_pending.

File: src/core/tcp_in.c [hunk 3]
```diff
@@ -617,6 +618,231 @@ tcp_input_delayed_close(struct tcp_pcb *pcb)
...
+        optlen = tcp_get_next_optbyte();
+        if (optlen < 2)
+          break;
+        tcp_optidx += optlen - 2;
+        break;
+    }
+  }
+
+  cookie = mss_idx;
+  cookie = (cookie << 24) | tcp_syncookie_calc();
```
Comment (wjhun) [inline src/core/tcp_in.c]: The double assignment isn't necessary; mss_idx can just be a u32.

File: src/core/tcp_in.c [hunk 4]
```diff
@@ -936,37 +1144,18 @@ tcp_process(struct tcp_pcb *pcb)
...
-            /* Call the accept function. */
-            TCP_EVENT_ACCEPT(pcb->listener, pcb, pcb->callback_arg, ERR_OK, err);
-          }
-          if (err != ERR_OK) {
-            /* If the accept function returns with an error, we abort
-             * the connection. */
-            /* Already aborted? */
-            if (err != ERR_ABRT) {
+#if TCP_LISTEN_BACKLOG
+            pcb->listener->syn_pending--;
```
Comment (wjhun) [inline src/core/tcp_in.c:1094]: assert here too

**Review by wjhun — COMMENTED** [review]
When comparing this implementation to the one in Linux, it appears that they differ in that the latter adds the client initial client sequence number when formulating a SYN cookie. When looking into this, I found this explanation from Robert Love of why this is being done:

https://www.quora.com/Why-does-Linuxs-SYN-cookie-implementation-use-the-initial-client-sequence-number-in-generating-the-cookie?share=1

I am loosely paraphrasing here, but it sounds like using a computed hash without adding the client ISN could potentially result in a cookie that is less than a previously-used cookie for the same tuple [Edit: It sounds like the goal is to reflect the spacing in ISNs that is provided by the client stack.] According to RFC6528, section 2 (https://datatracker.ietf.org/doc/html/rfc6528#section-2), some TCP stacks assume the property that ISNs are monotonically increasing in order to implement certain heuristics used when determining to create a new connection that may supersede a previous incarnation in TIME-WAIT state, and so violating this property could break such heuristics.

You might want to consider having this implementation follow in suit to avoid inter-operability issues with such TCP stacks that employ said heuristics.

File: src/core/tcp.c (hunk 1 above)
Comment (francescolavra) [reply src/core/tcp.c]: Fixed

File: src/core/tcp.c (hunk 2 above)
Comment (francescolavra) [reply src/core/tcp.c:333]: Added assert(pcb->listener->syn_pending != 0), both here and in the other place where syn_pending is decremented

File: src/core/tcp_in.c (hunk 3 above)
Comment (francescolavra) [reply src/core/tcp_in.c]: Now mss_idx is an u32_t, and the double assignment has been removed.

**Comment by francescolavra** [comment]
> You might want to consider having this implementation follow in suit to avoid inter-operability issues with such TCP stacks that employ said heuristics.

Makes sense, I changed the code that sends the syncookie so that the client initial sequence number is added to the computed cookie, and in tcp_syncookie_recv() the sequence number is removed before validating the cookie. In this function, I also fixed the comparison between the computed cookie and the received ACK number, which otherwise would fail when the `t` value is non-zero; now the 0xf8ffffff bitmask is applied when comparing the values, so that any non-zero `t` value is taken into account.

**Review by wjhun — APPROVED** [review]
LGTM

**Comment by francescolavra** [comment]
When doing some stress testing, I found an issue where tcp_receive() was accessing invalid data, and the reason was that an input pbuf that establishes a connection initiated via a syncookie was not being fully processed by tcp_input(). Now this is fixed by making tcp_listen_input() return the newly created pcb when a received syncookie is validated successfully, so that tcp_input() can continue processing the input pbuf and then call safely tcp_process(), which in turn calls tcp_receive() when handling the SYN_RCVD pcb state.

**Pull request merged**